File: common/src/main/java/com/technicjelle/bluemapofflineplayermarkers/Bukkit2Forge.java
package com.technicjelle.bluemapofflineplayermarkers;

import com.technicjelle.bluemapofflineplayermarkers.struct.OfflinePlayer;
import net.minecraft.CrashReport;
import net.minecraft.CrashReportCategory;
import net.minecraft.ReportedException;
import net.minecraft.nbt.*;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.util.FastBufferedInputStream;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.storage.LevelResource;
import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

public final class Bukkit2Forge {


    public static List<OfflinePlayer> getOfflinePlayers(MinecraftServer server) {

        List<OfflinePlayer> offlinePlayers = new ArrayList<>();
        List<String> playersOnServer = server.getPlayerList().getPlayers().stream().map(Entity::getStringUUID).toList();
        File[] files = new File(server.getWorldPath(LevelResource.PLAYER_DATA_DIR).toFile(), BlueMapOfflinePlayerMarkers.MOD_ID).listFiles();

        if (files == null)
                files = new File[0];
        else
            files = Arrays.stream(files).filter(playerFile -> !playersOnServer.contains(playerFile.getName().replaceFirst("(\\.dat)$", ""))).toArray(File[]::new);

        for (File file : files) {
            offlinePlayers.add(getDataFromFiles(file));
        }

        return offlinePlayers;

    }

    public static OfflinePlayer toOfflinePlayer(ServerPlayer player) {
        return getDataFromFiles(
                new File(
                        player.getServer().getWorldPath(LevelResource.PLAYER_DATA_DIR).toFile(),
                        BlueMapOfflinePlayerMarkers.MOD_ID + File.separator + player.getStringUUID() + ".dat"
                )
        );
    }


    public static OfflinePlayer getDataFromFiles(File playerData) {

        try (FileInputStream stream = new FileInputStream(playerData)) {

            CompoundTag nbt = readCompressedNbt(stream);

            return new OfflinePlayer(
                    UUID.fromString(playerData.getName().replaceFirst("(\\.dat)$", "")),
                    nbt.getString("username"),
                    nbt.getLong("lastOnline"),
                    nbt.getList("position", 6).stream().map( tag -> (tag instanceof DoubleTag doubleTag) ? doubleTag.getAsDouble() : 0.0d).toList(),
                    nbt.getString("dimension"),
                    nbt.getInt("gameMode")
            );

        } catch (Exception ignored) {} // EXCEPTION IGNORED FOR NOW

        // It should already be broken by the time it gets here
        return null;

    }

    public static void writePlayerNbt(ServerPlayer player) {

        File offlinePlayers = new File(player.getServer().getWorldPath(LevelResource.PLAYER_DATA_DIR).toFile(), BlueMapOfflinePlayerMarkers.MOD_ID);

        if (!offlinePlayers.exists()) {
            offlinePlayers.mkdir();
        }

        if (!offlinePlayers.exists())
            return ; // ERROR

        CompoundTag nbt = new CompoundTag();

        nbt.putString("username", player.getName().getString());
        nbt.putLong("lastOnline", System.currentTimeMillis());
        nbt.putString("dimension", player.level().dimension().location().toString());
        nbt.putInt("gameMode", player.gameMode.getGameModeForPlayer().getId());

        ListTag position = new ListTag();

        position.add(DoubleTag.valueOf(player.position().x));
        position.add(DoubleTag.valueOf(player.position().y));
        position.add(DoubleTag.valueOf(player.position().z));

        nbt.put("position", position);

        try {
            writeCompressedNbt(nbt, new File(offlinePlayers, player.getStringUUID() + ".dat"));
        } catch (IOException ignored) {} // EXCEPTION IGNORED FOR NOW

    }

    public static void writeCompressedNbt(CompoundTag tag, File file) throws IOException {

        try (FileOutputStream outputStream = new FileOutputStream(file)) {
            try (DataOutputStream dataOutputStream = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(outputStream)))) {
                writeTag(tag, dataOutputStream);
            }
        }

    }

    public static void writeTag(@Nullable Tag tag, DataOutput dataOutput) throws IOException {

        if(tag == null)
            return ;

        dataOutput.write(tag.getId());

        if (tag.getId() != 0) {
            dataOutput.writeUTF("");
            tag.write(dataOutput);
        }

    }

    public static CompoundTag readCompressedNbt(@Nullable InputStream stream) throws IOException {

        if (stream == null)
            return null; // May break things

        return readNbt(new DataInputStream(new FastBufferedInputStream(new GZIPInputStream(stream))), NbtAccounter.unlimitedHeap());

    }

    public static CompoundTag readNbt(@Nullable DataInput dataInput, @Nullable NbtAccounter nbtAccounter) throws IOException {

        if (dataInput == null || nbtAccounter == null)
            return null;

        Tag tag = readUnnamedTag(dataInput, nbtAccounter);

        if (tag instanceof CompoundTag compoundTag) {
            return compoundTag;
        } else {
            throw new IOException("Root tag must be a named compound tag");
        }

    }

    private static Tag readUnnamedTag(DataInput dataInput, NbtAccounter nbtAccounter) throws IOException {

        byte b0 = dataInput.readByte();
        nbtAccounter.accountBytes(1L);

        if (b0 == 0)
            return EndTag.INSTANCE;

        readUTF(nbtAccounter, dataInput.readUTF());
        nbtAccounter.accountBytes(4L);

        try {
            return TagTypes.getType(b0).load(dataInput, nbtAccounter);
        } catch (IOException exception) {
            CrashReport crashreport = CrashReport.forThrowable(exception, "Loading NBT data");
            CrashReportCategory category = crashreport.addCategory("NBT Tag");
            category.setDetail("Tag type", b0);
            throw new ReportedException(crashreport);
        }
    }

    @Nullable
    public static String readUTF(NbtAccounter nbtAccounter, @Nullable String data) {

        nbtAccounter.accountBytes(16L);

        if (data == null)
            return data;

        int len = data.length();
        int utfLen = 0;

        for (int i = 0; i < len; i++) {

            int c = data.charAt(i);

            if (c >= 1 && c <= 127) {
                utfLen++;
            } else if (c > 2047) {
                utfLen += 3;
            } else {
                utfLen += 2;
            }

        }

        nbtAccounter.accountBytes((8L * utfLen));

        return data;

    }

}

File: src/main/java/com/technicjelle/bluemapofflineplayermarkers/Bukkit2Forge.kt
package com.technicjelle.bluemapofflineplayermarkers

import net.minecraft.CrashReport
import net.minecraft.ReportedException
import net.minecraft.nbt.*
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerPlayer
import net.minecraft.util.FastBufferedInputStream
import net.minecraft.world.level.storage.LevelResource
import java.io.*
import java.util.*
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

fun MinecraftServer.getOfflinePlayers(): List<OfflinePlayer> = mutableListOf<OfflinePlayer>().apply {

    val playersOnServer: List<String> = playerList.players.map { it.stringUUID }

    val offlineFiles: List<File> =
        File(getWorldPath(LevelResource.PLAYER_DATA_DIR).toFile(), BlueMapOfflinePlayerMarkers.MOD_ID).listFiles()?.filter {
            !playersOnServer.contains(it.name.removeSuffix(".dat"))
        } ?: emptyList()

    if (offlineFiles.isEmpty()) return@apply

    offlineFiles.forEach { file ->
        add(getDataFromFiles(file))
    }
}

fun ServerPlayer.toOfflinePlayer(): OfflinePlayer = getDataFromFiles(
    File(
        server.getWorldPath(LevelResource.PLAYER_DATA_DIR).toFile(),
        "${BlueMapOfflinePlayerMarkers.MOD_ID}${File.separator}${stringUUID}.dat"
    )
)


fun getDataFromFiles(offlineData: File): OfflinePlayer {
    FileInputStream(offlineData).use { stream ->
        readCompressedNbt(stream).let { nbt ->
            return OfflinePlayer(
                UUID.fromString(offlineData.name.removeSuffix(".dat")),
                nbt.getString("username"),
                nbt.getLong("lastOnline"),
                nbt.getList("position", 6).map { tag -> (tag as DoubleTag).asDouble },
                nbt.getString("dimension"),
                nbt.getInt("gameMode")
            )
        }
    }
}

fun ServerPlayer.writePlayerNbt() {
    File(server.getWorldPath(LevelResource.PLAYER_DATA_DIR).toFile(), BlueMapOfflinePlayerMarkers.MOD_ID).let {
        if (!it.exists()) it.mkdir()
        if (!it.exists()) it.createNewFile()
        writeCompressedNbt(CompoundTag().apply {
            putString("username", name.string)
            putLong("lastOnline", System.currentTimeMillis())
            put("position", ListTag().apply {
                add(DoubleTag.valueOf(position().x))
                add(DoubleTag.valueOf(position().y))
                add(DoubleTag.valueOf(position().z))
            })
            putString("dimension", level().dimension().location().toString())
            putInt("gameMode", gameMode.gameModeForPlayer.id)
        }, File(it, "$stringUUID.dat"))
    }
}


@Throws(IOException::class)
fun writeCompressedNbt(tag: CompoundTag, file: File) {
    FileOutputStream(file).use { outputstream ->
        DataOutputStream(BufferedOutputStream(GZIPOutputStream(outputstream))).use { dataoutputstream ->
            writeTag(tag, dataoutputstream)
        }
    }
}

@Throws(IOException::class)
fun writeTag(tag: Tag?, dataOutput: DataOutput) {

    if(tag == null) return

    dataOutput.writeByte(tag.id.toInt())
    if (tag.id.toInt() != 0) {
        dataOutput.writeUTF("")
        tag.write(dataOutput)
    }
}

@Throws(IOException::class)
fun readCompressedNbt(stream: InputStream?): CompoundTag {

    try {
        return readNbt(DataInputStream(FastBufferedInputStream(GZIPInputStream(stream))),
            NbtAccounter.unlimitedHeap()
        )
    }catch(exception: IOException) {
        throw exception
    }

}


@Throws(IOException::class)
fun readNbt(dataInput: DataInput?, nbtAccounter: NbtAccounter?): CompoundTag {
    val tag = dataInput?.let { nbtAccounter?.let { it1 -> readUnnamedTag(it, it1) } }
    if (tag is CompoundTag) {
        return tag
    } else {
        throw IOException("Root tag must be a named compound tag")
    }
}


@Throws(IOException::class)
private fun readUnnamedTag(dataInput: DataInput, nbtAccounter: NbtAccounter): Any? {
    val b0 = dataInput.readByte()
    nbtAccounter.accountBytes(1L)
    return if (b0.toInt() == 0) {
        EndTag.INSTANCE
    } else {
        readUTF(nbtAccounter, dataInput.readUTF())
        nbtAccounter.accountBytes(4L)
        try {
            TagTypes.getType(b0.toInt()).load(dataInput, nbtAccounter)
        } catch (ioexception: IOException) {
            val crashreport = CrashReport.forThrowable(ioexception, "Loading NBT data")
            val crashreportcategory = crashreport.addCategory("NBT Tag")
            crashreportcategory.setDetail("Tag type", b0)
            throw ReportedException(crashreport)
        }
    }
}

fun readUTF(nbtAccounter: NbtAccounter, data: String?): String? {
    nbtAccounter.accountBytes(16L)
    if (data == null) {
        return data
    } else {
        val len = data.length
        var utflen = 0

        for (i in 0 until len) {
            val c = data[i].code
            if (c in 1..127) {
                ++utflen
            } else if (c > 2047) {
                utflen += 3
            } else {
                utflen += 2
            }
        }

        nbtAccounter.accountBytes((8 * utflen).toLong())
        return data
    }
}